manifests/init.pp

```diff
@@ -0,0 +1,128 @@
+# @summary This is a `firewalld` profile that sets "safe" defaults as is usual in SIMP modules
+#
+# If you want to override any element not present in the `firewalld` class
+# resource below then you should use Hiera directly on the `firewalld` class.
+#
+# @param enable
+#   Activate the firewalld management capabilties.
+#
+#   * The class will not be enabled if firewalld is not detected on the remote
+#     system. This can be overridden by setting this option to `true`
+#     explicitly in Hiera.
+#
+# @param complete_reload
+#   The current firewalld module has the capability to perform a complete reload
+#   of firewalld which breaks any existing connections. This is extremely
+#   dangerous and this class overrides and disables this capability by default.
+#
+#   * Set to ``true`` to re-enable this capability.
+#
+# @param lockdown
+#   Set ``firewalld`` in ``lockdown`` mode which disallows manipulation by
+#   applications.
+#
+#   * This makes sense to do by default since puppet is meant to be
+#     authoritative on the system.
+#
+# @param default_zone
+#   The 'default zone' to set on the system.
+#
+#   This is set to ``99_simp`` so that regular, alternative, zone manipulation
+#   can occur without interference.
+#
+#   **IMPORTANT:** If this is set to anything besides ``99_simp``, all rules in
+#   this module will **NOT** apply to the default zone! This module is set to
+#   only populate ``99_simp`` zone rules.
+#
+# @param log_denied
+#   What types of logs to process for denied packets.
+#
+#   @see LogDenied in firewalld.conf(5)
+#
+# @param firewall_backend
+#   Allows you to set the backend that firewalld will use.
+#
+#   * Currently set to 'iptables' due to bugs in nftables
+#
+# @param enable_tidy
+#   Enable the ``Tidy`` resources that help keep the system clean from cruft
+#
+# @param tidy_dirs
+#   The directories to target for tidying
+#
+# @param tidy_prefix
+#   The name match to use for tidying files
+#
+# @param tidy_minutes
+#   Number of **minutes** to consider a configuration file 'stale' for the
+#   purposes of tidying.
+#
+# @param simp_zone_interfaces
+#   The network interfaces to which the underlying 99_simp zone should apply
+#
+# @param simp_zone_target
+#   The default target for the 99_simp zone
+#
+class simp_firewalld (
+  Boolean                                              $enable               = 'firewalld' in pick($facts['simplib__firewalls'], 'none'),
```

Comment:
Should consider `$simp_options::firewall` when available
 (note that  `$simp_options::firewall` is a "stroolean" that can be `true`, `false`, or `"firewalld"`: https://github.com/simp/pupmod-simp-simp_options/blob/master/manifests/init.pp#L98`)

Reply:
May have a point here; I could see this being a cart-before-the-horse scenario as well. If we have a system that does not have firewalld installed anymore (i.e. EL7 with just iptables, firewalld removed) we would want to potentially set this to enabled without having to double-override if `simp_options::firewall` is already set.

Reply:
@ayohrling: I think scenarios like that boil down to trusting that the classification of `simp_firewalld` is intentional.

If a system no longer has firewalld installed and is using iptables instead, the node should not be classified with `simp_firewalld` and this parameter (along with any default logic) will not be invoked.

Reply:
Hrm...I guess this is less `enable` and more `manage`.

The flag was there to make sure that the class wasn't managed if `firewalld` wasn't present on the host at all. It alleviates the need to mess around with exact version matching and everything else.

* If it's there => Manage it by default
* If it's not there => Ignore it
* Manual => disable if you want to

Reply:
Same as simp_options below theory-wise. If we give users/administrators tunables to globally affect the system, but then we require them to not only set those tunables but also have to set an additional setting manually in `simp_firewalld::enable: true` to actually enable firewalld from a previous instance, we're not creating a cohesive product. It's not "hard" at all, of course, but it just seems non-cohesive to have to set multiple hiera values but call one a global catalyst...

Reply:
@ayohrling This is fundamentally the issue around having independent modules and a cohesive product. We *had* a cohesive product but people wanted to be able to use the modules independently so it got a bit messy. Not much of an "easy" button around it unfortunately.

Reply:
Yeah, I mean again case-by-case, in this case it's a simp_* wrapper of firewalld; implying that someone could easily use puppet-firewalld separately without issue; a SIMP wrapper implies coupling.

Reply:
@ayohrling Would you prefer that this just be a toggle on the internal logic? It makes no sense to try to manage firewalld if there is no firewalld to manage. It's a bit tricky since, theoretically, this is a classification issue but I was trying to avoid making firewalld mandatory and/or making it difficult to mix in.

Reply:
Yeah, I think I'm leaning towards this being a toggle based on the simp_options logic as far as the param; perhaps we could have the code inside the class check firewalld support (i.e. metadata of module or a custom fact) instead of the pick on simplib::firewalls since that implies installed firewalls... at the end of the day it's not the worst, I just see a scenario where an admin has an older system with iptables and wants to move their systems to firewalld, sets simp_options, and nothing changes since this returns 'false' from the pick. Conversely, I can see an issue where we set firewalld, and some EL6 nodes then can't install it because this is enabled. So we should have some "supportability" check, but err towards newer systems (i.e. EL6 is going away)

Reply:
I think @op-ct was the one weighing in previously here, would also ask for secondary opinion?

Reply:
So, the `simplib__firewalls` fact detects *available* firewalls, not active/installed ones. This means that the class won't activate if `firewalld` isn't on the system at all and will otherwise.

I went for the "safe" version of things by doing this. Default installations of EL7 and EL8 have `firewalld` installed and the system owner would have had to specifically not installed `firewalld` for some reason to make this not true.

But, I added it as a class parameter so that they could force the issue if they felt the need to do so and, likewise, force not managing firewalld if they felt the need to do so.

I think that covers all of the logic behind the current design.

Reply:
I'm good w/ that.

Reply:
As a followup to the discussion over this line: the recent "Stroolean" `firewalld` addition to `$simp_options::firewall` really confused the purpose of this otherwise-simple global catalyst (here and elsewhere), so I've put in an issue to revert it and handle provider hinting elsewhere (probably in iptables): [SIMP-7818]